### PR TITLE
op-chain-ops: wipe predeploy storage

### PR DESCRIPTION
**Pull request by tynes**
**Description**

Makes the spec for the migration process much more simple by wiping the storage of the predeploys expect for special cases. The contracts that do not have their storage wiped are:

- GovernanceToken
- WETH9
- LegacyMessagePasser
- LegacyERC20ETH
- DeployerWhitelist

### CODE REVIEW

**Review by mslipper — CHANGES_REQUESTED** [review]
I'm generally in favor of erasing storage to simplify the migration. However, I have some concerns with the implementation here that I'd like to address before merging.

- Have we audited all of the predeploys to ensure that we're not accidentally erasing necessary state, or if we are that the state gets reset in subsequent migration operations? The concern I have is that we un-initialize something that used to be initialized, leading to something like the Nomad hack. Would love to see some bullets talking about which state gets erased for each predeploy, and whether that state needs to be reset later on in the migration.
- Have you been able to run this on a datadir, and confirmed that all the checks pass? I can set this up for you if it'd be helpful. Looking at the code here, I'm 80% sure that the storage for OVM_ETH will be zeroed out as well.

**Comment by tynes** [comment]
Converting this to a draft until bullet points in https://github.com/ethereum-optimism/optimism/pull/4600#pullrequestreview-1236807650 are addressed

Perhaps another approach at this would be to zero out specific storage slots instead of zeroing them all out. It adds some complexity to the spec but there is less risk of an unknown storage slot being reset. @smartcontracts requested this while writing the migration spec

I did push a commit that skips resetting the storage of OVM_ETH. I haven't ran this on a datadir yet

**Comment by mslipper** [comment]
Cool, let's jam tomorrow. I have a script you can use that will give you a clean datadir to run on. Should make the feedback loop for running the migration much faster.

**Comment by smartcontracts** [comment]
> Have we audited all of the predeploys to ensure that we're not accidentally erasing necessary state, or if we are that the state gets reset in subsequent migration operations?

I've manually reviewed the predeploys and can confirm that only the contracts listed in `contracts/src/predeploys.ts` contain any code or state. We have also carefully analyzed every predeploy and can confirm that wiping everything EXCEPT LegacyMessagePasser, DeployerWhitelist, GovernanceToken, and LegacyERC20ETH is OK.

**Comment by tynes** [comment]
> > Have we audited all of the predeploys to ensure that we're not accidentally erasing necessary state, or if we are that the state gets reset in subsequent migration operations?
> 
> I've manually reviewed the predeploys and can confirm that only the contracts listed in `contracts/src/predeploys.ts` contain any code or state. We have also carefully analyzed every predeploy and can confirm that wiping everything EXCEPT LegacyMessagePasser, DeployerWhitelist, GovernanceToken, and LegacyERC20ETH is OK.

I believe we also need WETH9 in that list as well?

File: op-chain-ops/genesis/setters.go
Comment (smartcontracts) [inline op-chain-ops/genesis/setters.go:83]: In the interest of speed, let's do `db.CreateAccount` instead and call the function WipePredeploy

Comment (tynes) [reply op-chain-ops/genesis/setters.go]: Updated to this suggestion

**Comment by tynes** [comment]
I also added in the post check code so that we can have coverage that the storage has been wiped. There is a todo for building a map of the number of expected storage slots per contract and then asserting that we have exactly the correct number of storage slots

**Comment by tynes** [comment]
I've changed the base onto https://github.com/ethereum-optimism/optimism/pull/4605, going to need to fix a merge conflict

File: op-chain-ops/genesis/check.go
Comment (protolambda) [inline op-chain-ops/genesis/check.go:265]: Old todo?

File: op-chain-ops/genesis/check.go
Comment (mslipper) [inline op-chain-ops/genesis/check.go:264]: Can we also check that the storage root is the empty storage root for those contracts that have zero storage slots?

**Review by mslipper — APPROVED** [review]
Approving pending answer to my comment above.

File: op-chain-ops/genesis/check.go
Comment (smartcontracts) [reply op-chain-ops/genesis/check.go:264]: I think this is guaranteed implicitly

File: op-chain-ops/genesis/check.go
Comment (tynes) [inline op-chain-ops/genesis/check.go:40]: Each of these should should `+ 2` for admin + impl eip 1967 storage slots being set, if they are behind a proxy

File: op-chain-ops/genesis/check.go
Comment (tynes) [inline op-chain-ops/genesis/check.go:41]: slight nit to have a comment here describing this and saying it accounts for eip 1967 but lgtm otherwise

**Comment by tynes** [comment]
`foundryup` failed causing devnet + genesis contracts to fail, i think the api is timing out

**Comment by tynes** [comment]
I can't approve but looks good to me

**Pull request merged**
